File: android/src/main/java/expo/modules/smartcamera/MLKitFaceDetector.kt
package expo.modules.smartcamera

import android.graphics.Bitmap
import android.graphics.PointF
import android.graphics.Rect
import com.google.mlkit.vision.common.InputImage
import com.google.mlkit.vision.face.Face
import com.google.mlkit.vision.face.FaceContour
import com.google.mlkit.vision.face.FaceDetection
import com.google.mlkit.vision.face.FaceDetector
import com.google.mlkit.vision.face.FaceDetectorOptions
import com.google.mlkit.vision.face.FaceLandmark
import kotlinx.coroutines.suspendCancellableCoroutine
import kotlin.coroutines.resume
import kotlin.coroutines.resumeWithException

class MLKitFaceDetector {
    // MARK: - Properties
    
    private var detector: FaceDetector? = null
    private var currentOptions: FaceDetectionOptions = FaceDetectionOptions()
    
    // MARK: - Configuration
    
    fun updateOptions(options: FaceDetectionOptions) {
        if (options != currentOptions) {
            currentOptions = options
            detector?.close()
            detector = createDetector(options)
        }
    }
    
    private fun createDetector(options: FaceDetectionOptions): FaceDetector {
        val builder = FaceDetectorOptions.Builder()
        
        // Performance mode
        when (options.performanceMode) {
            "fast" -> builder.setPerformanceMode(FaceDetectorOptions.PERFORMANCE_MODE_FAST)
            "accurate" -> builder.setPerformanceMode(FaceDetectorOptions.PERFORMANCE_MODE_ACCURATE)
        }
        
        // Landmark mode
        when (options.landmarkMode) {
            "all" -> builder.setLandmarkMode(FaceDetectorOptions.LANDMARK_MODE_ALL)
            else -> builder.setLandmarkMode(FaceDetectorOptions.LANDMARK_MODE_NONE)
        }
        
        // Contour mode
        when (options.contourMode) {
            "all" -> builder.setContourMode(FaceDetectorOptions.CONTOUR_MODE_ALL)
            else -> builder.setContourMode(FaceDetectorOptions.CONTOUR_MODE_NONE)
        }
        
        // Classification mode
        when (options.classificationMode) {
            "all" -> builder.setClassificationMode(FaceDetectorOptions.CLASSIFICATION_MODE_ALL)
            else -> builder.setClassificationMode(FaceDetectorOptions.CLASSIFICATION_MODE_NONE)
        }
        
        // Min face size
        builder.setMinFaceSize(options.minFaceSize.toFloat())
        
        // Tracking
        if (options.trackingEnabled) {
            builder.enableTracking()
        }
        
        return FaceDetection.getClient(builder.build())
    }
    
    // MARK: - Detection
    
    suspend fun detectFaces(bitmap: Bitmap): List<DetectedFace> {
        val faceDetector = detector ?: createDetector(currentOptions).also { detector = it }
        val inputImage = InputImage.fromBitmap(bitmap, 0)
        
        return suspendCancellableCoroutine { continuation ->
            faceDetector.process(inputImage)
                .addOnSuccessListener { faces ->
                    val detectedFaces = faces.mapIndexed { index, face ->
                        convertToDetectedFace(face, index)
                    }
                    continuation.resume(detectedFaces)
                }
                .addOnFailureListener { exception ->
                    continuation.resumeWithException(exception)
                }
        }
    }
    
    // MARK: - Conversion
    
    private fun convertToDetectedFace(face: Face, index: Int): DetectedFace {
        val bounds = face.boundingBox
        
        return DetectedFace(
            bounds = FaceBounds(
                x = bounds.left.toDouble(),
                y = bounds.top.toDouble(),
                width = bounds.width().toDouble(),
                height = bounds.height().toDouble()
            ),
            landmarks = if (currentOptions.landmarkMode == "all") extractLandmarks(face) else null,
            contours = if (currentOptions.contourMode == "all") extractContours(face) else null,
            smilingProbability = face.smilingProbability?.toDouble(),
            leftEyeOpenProbability = face.leftEyeOpenProbability?.toDouble(),
            rightEyeOpenProbability = face.rightEyeOpenProbability?.toDouble(),
            trackingId = if (currentOptions.trackingEnabled) face.trackingId else null,
            headEulerAngleX = face.headEulerAngleX.toDouble(),
            headEulerAngleY = face.headEulerAngleY.toDouble(),
            headEulerAngleZ = face.headEulerAngleZ.toDouble()
        )
    }
    
    private fun extractLandmarks(face: Face): FaceLandmarksData {
        return FaceLandmarksData(
            leftEye = face.getLandmark(FaceLandmark.LEFT_EYE)?.position?.toPointData(),
            rightEye = face.getLandmark(FaceLandmark.RIGHT_EYE)?.position?.toPointData(),
            leftEar = face.getLandmark(FaceLandmark.LEFT_EAR)?.position?.toPointData(),
            rightEar = face.getLandmark(FaceLandmark.RIGHT_EAR)?.position?.toPointData(),
            leftCheek = face.getLandmark(FaceLandmark.LEFT_CHEEK)?.position?.toPointData(),
            rightCheek = face.getLandmark(FaceLandmark.RIGHT_CHEEK)?.position?.toPointData(),
            noseBase = face.getLandmark(FaceLandmark.NOSE_BASE)?.position?.toPointData(),
            mouthLeft = face.getLandmark(FaceLandmark.MOUTH_LEFT)?.position?.toPointData(),
            mouthRight = face.getLandmark(FaceLandmark.MOUTH_RIGHT)?.position?.toPointData(),
            mouthBottom = face.getLandmark(FaceLandmark.MOUTH_BOTTOM)?.position?.toPointData()
        )
    }
    
    private fun extractContours(face: Face): FaceContoursData {
        return FaceContoursData(
            face = face.getContour(FaceContour.FACE)?.points?.map { it.toPointData() },
            leftEyebrowTop = face.getContour(FaceContour.LEFT_EYEBROW_TOP)?.points?.map { it.toPointData() },
            leftEyebrowBottom = face.getContour(FaceContour.LEFT_EYEBROW_BOTTOM)?.points?.map { it.toPointData() },
            rightEyebrowTop = face.getContour(FaceContour.RIGHT_EYEBROW_TOP)?.points?.map { it.toPointData() },
            rightEyebrowBottom = face.getContour(FaceContour.RIGHT_EYEBROW_BOTTOM)?.points?.map { it.toPointData() },
            leftEye = face.getContour(FaceContour.LEFT_EYE)?.points?.map { it.toPointData() },
            rightEye = face.getContour(FaceContour.RIGHT_EYE)?.points?.map { it.toPointData() },
            upperLipTop = face.getContour(FaceContour.UPPER_LIP_TOP)?.points?.map { it.toPointData() },
            upperLipBottom = face.getContour(FaceContour.UPPER_LIP_BOTTOM)?.points?.map { it.toPointData() },
            lowerLipTop = face.getContour(FaceContour.LOWER_LIP_TOP)?.points?.map { it.toPointData() },
            lowerLipBottom = face.getContour(FaceContour.LOWER_LIP_BOTTOM)?.points?.map { it.toPointData() },
            noseBridge = face.getContour(FaceContour.NOSE_BRIDGE)?.points?.map { it.toPointData() },
            noseBottom = face.getContour(FaceContour.NOSE_BOTTOM)?.points?.map { it.toPointData() }
        )
    }
    
    // MARK: - Cleanup
    
    fun close() {
        detector?.close()
        detector = null
    }
}

// MARK: - Extensions

private fun PointF.toPointData(): PointData = PointData(x.toDouble(), y.toDouble())

// MARK: - Data Classes

data class DetectedFace(
    val bounds: FaceBounds,
    val landmarks: FaceLandmarksData? = null,
    val contours: FaceContoursData? = null,
    val smilingProbability: Double? = null,
    val leftEyeOpenProbability: Double? = null,
    val rightEyeOpenProbability: Double? = null,
    val trackingId: Int? = null,
    val headEulerAngleX: Double? = null,
    val headEulerAngleY: Double? = null,
    val headEulerAngleZ: Double? = null
) {
    fun toMap(): Map<String, Any> {
        val map = mutableMapOf<String, Any>(
            "bounds" to mapOf(
                "x" to bounds.x,
                "y" to bounds.y,
                "width" to bounds.width,
                "height" to bounds.height
            )
        )
        
        landmarks?.let { map["landmarks"] = it.toMap() }
        contours?.let { map["contours"] = it.toMap() }
        smilingProbability?.let { map["smilingProbability"] = it }
        leftEyeOpenProbability?.let { map["leftEyeOpenProbability"] = it }
        rightEyeOpenProbability?.let { map["rightEyeOpenProbability"] = it }
        trackingId?.let { map["trackingId"] = it }
        headEulerAngleX?.let { map["headEulerAngleX"] = it }
        headEulerAngleY?.let { map["headEulerAngleY"] = it }
        headEulerAngleZ?.let { map["headEulerAngleZ"] = it }
        
        return map
    }
}

data class FaceBounds(
    val x: Double,
    val y: Double,
    val width: Double,
    val height: Double
)

data class PointData(
    val x: Double,
    val y: Double
) {
    fun toMap(): Map<String, Any> = mapOf("x" to x, "y" to y)
}

data class FaceLandmarksData(
    val leftEye: PointData? = null,
    val rightEye: PointData? = null,
    val leftEar: PointData? = null,
    val rightEar: PointData? = null,
    val leftCheek: PointData? = null,
    val rightCheek: PointData? = null,
    val noseBase: PointData? = null,
    val mouthLeft: PointData? = null,
    val mouthRight: PointData? = null,
    val mouthBottom: PointData? = null
) {
    fun toMap(): Map<String, Any> {
        val map = mutableMapOf<String, Any>()
        leftEye?.let { map["leftEye"] = it.toMap() }
        rightEye?.let { map["rightEye"] = it.toMap() }
        leftEar?.let { map["leftEar"] = it.toMap() }
        rightEar?.let { map["rightEar"] = it.toMap() }
        leftCheek?.let { map["leftCheek"] = it.toMap() }
        rightCheek?.let { map["rightCheek"] = it.toMap() }
        noseBase?.let { map["noseBase"] = it.toMap() }
        mouthLeft?.let { map["mouthLeft"] = it.toMap() }
        mouthRight?.let { map["mouthRight"] = it.toMap() }
        mouthBottom?.let { map["mouthBottom"] = it.toMap() }
        return map
    }
}

data class FaceContoursData(
    val face: List<PointData>? = null,
    val leftEyebrowTop: List<PointData>? = null,
    val leftEyebrowBottom: List<PointData>? = null,
    val rightEyebrowTop: List<PointData>? = null,
    val rightEyebrowBottom: List<PointData>? = null,
    val leftEye: List<PointData>? = null,
    val rightEye: List<PointData>? = null,
    val upperLipTop: List<PointData>? = null,
    val upperLipBottom: List<PointData>? = null,
    val lowerLipTop: List<PointData>? = null,
    val lowerLipBottom: List<PointData>? = null,
    val noseBridge: List<PointData>? = null,
    val noseBottom: List<PointData>? = null
) {
    fun toMap(): Map<String, Any> {
        val map = mutableMapOf<String, Any>()
        face?.let { map["face"] = it.map { p -> p.toMap() } }
        leftEyebrowTop?.let { map["leftEyebrowTop"] = it.map { p -> p.toMap() } }
        leftEyebrowBottom?.let { map["leftEyebrowBottom"] = it.map { p -> p.toMap() } }
        rightEyebrowTop?.let { map["rightEyebrowTop"] = it.map { p -> p.toMap() } }
        rightEyebrowBottom?.let { map["rightEyebrowBottom"] = it.map { p -> p.toMap() } }
        leftEye?.let { map["leftEye"] = it.map { p -> p.toMap() } }
        rightEye?.let { map["rightEye"] = it.map { p -> p.toMap() } }
        upperLipTop?.let { map["upperLipTop"] = it.map { p -> p.toMap() } }
        upperLipBottom?.let { map["upperLipBottom"] = it.map { p -> p.toMap() } }
        lowerLipTop?.let { map["lowerLipTop"] = it.map { p -> p.toMap() } }
        lowerLipBottom?.let { map["lowerLipBottom"] = it.map { p -> p.toMap() } }
        noseBridge?.let { map["noseBridge"] = it.map { p -> p.toMap() } }
        noseBottom?.let { map["noseBottom"] = it.map { p -> p.toMap() } }
        return map
    }
}

File: ios/MLKitFaceDetector.swift
import UIKit
import Vision

// Note: This is a placeholder implementation.
// In production, you would use GoogleMLKit/FaceDetection pod.
// For now, we use Apple's Vision framework as a fallback.

class MLKitFaceDetector {
    // MARK: - Properties
    
    private var performanceMode: String = "fast"
    private var landmarkMode: String = "none"
    private var contourMode: String = "none"
    private var classificationMode: String = "none"
    private var minFaceSize: Double = 0.15
    private var trackingEnabled: Bool = false
    
    // Vision request
    private lazy var faceDetectionRequest: VNDetectFaceLandmarksRequest = {
        let request = VNDetectFaceLandmarksRequest()
        return request
    }()
    
    // MARK: - Configuration
    
    func updateOptions(_ options: FaceDetectionOptions) {
        performanceMode = options.performanceMode
        landmarkMode = options.landmarkMode
        contourMode = options.contourMode
        classificationMode = options.classificationMode
        minFaceSize = options.minFaceSize
        trackingEnabled = options.trackingEnabled
    }
    
    // MARK: - Detection
    
    func detectFaces(in image: UIImage) async throws -> [DetectedFace] {
        guard let cgImage = image.cgImage else {
            throw NSError(domain: "MLKitFaceDetector", code: 1, userInfo: [
                NSLocalizedDescriptionKey: "Could not get CGImage from UIImage"
            ])
        }
        
        return try await withCheckedThrowingContinuation { continuation in
            let handler = VNImageRequestHandler(cgImage: cgImage, options: [:])
            
            let request = VNDetectFaceLandmarksRequest { [weak self] request, error in
                if let error = error {
                    continuation.resume(throwing: error)
                    return
                }
                
                guard let observations = request.results as? [VNFaceObservation] else {
                    continuation.resume(returning: [])
                    return
                }
                
                let faces = observations.enumerated().compactMap { index, observation in
                    self?.convertToDetectedFace(observation, index: index, imageSize: image.size)
                }
                
                continuation.resume(returning: faces)
            }
            
            do {
                try handler.perform([request])
            } catch {
                continuation.resume(throwing: error)
            }
        }
    }
    
    func detectFaces(in sampleBuffer: CMSampleBuffer) -> [DetectedFace] {
        guard let pixelBuffer = CMSampleBufferGetImageBuffer(sampleBuffer) else {
            return []
        }
        
        let handler = VNImageRequestHandler(cvPixelBuffer: pixelBuffer, options: [:])
        
        do {
            try handler.perform([faceDetectionRequest])
            
            guard let observations = faceDetectionRequest.results else {
                return []
            }
            
            let imageWidth = CVPixelBufferGetWidth(pixelBuffer)
            let imageHeight = CVPixelBufferGetHeight(pixelBuffer)
            let imageSize = CGSize(width: imageWidth, height: imageHeight)
            
            return observations.enumerated().compactMap { index, observation in
                convertToDetectedFace(observation, index: index, imageSize: imageSize)
            }
        } catch {
            print("[MLKitFaceDetector] Error detecting faces: \(error)")
            return []
        }
    }
    
    // MARK: - Conversion
    
    private func convertToDetectedFace(_ observation: VNFaceObservation, index: Int, imageSize: CGSize) -> DetectedFace {
        // Convert normalized coordinates to pixel coordinates
        let boundingBox = observation.boundingBox
        let x = boundingBox.origin.x * imageSize.width
        let y = (1 - boundingBox.origin.y - boundingBox.height) * imageSize.height
        let width = boundingBox.width * imageSize.width
        let height = boundingBox.height * imageSize.height
        
        var face = DetectedFace(
            bounds: FaceBounds(x: x, y: y, width: width, height: height),
            trackingId: trackingEnabled ? index : nil
        )
        
        // Add landmarks if requested
        if landmarkMode == "all", let landmarks = observation.landmarks {
            face.landmarks = extractLandmarks(landmarks, imageSize: imageSize)
        }
        
        // Add head pose
        if let yaw = observation.yaw?.doubleValue {
            face.headEulerAngleY = yaw * 180 / .pi
        }
        if let roll = observation.roll?.doubleValue {
            face.headEulerAngleZ = roll * 180 / .pi
        }
        
        return face
    }
    
    private func extractLandmarks(_ landmarks: VNFaceLandmarks2D, imageSize: CGSize) -> FaceLandmarksData {
        var data = FaceLandmarksData()
        
        if let leftEye = landmarks.leftEye {
            data.leftEye = averagePoint(leftEye.normalizedPoints, imageSize: imageSize)
        }
        if let rightEye = landmarks.rightEye {
            data.rightEye = averagePoint(rightEye.normalizedPoints, imageSize: imageSize)
        }
        if let nose = landmarks.nose {
            data.noseBase = averagePoint(nose.normalizedPoints, imageSize: imageSize)
        }
        if let outerLips = landmarks.outerLips {
            let points = outerLips.normalizedPoints
            if points.count > 0 {
                data.mouthLeft = convertPoint(points[0], imageSize: imageSize)
            }
            if points.count > points.count / 2 {
                data.mouthRight = convertPoint(points[points.count / 2], imageSize: imageSize)
            }
        }
        
        return data
    }
    
    private func averagePoint(_ points: [CGPoint], imageSize: CGSize) -> PointData {
        guard !points.isEmpty else {
            return PointData(x: 0, y: 0)
        }
        
        let sumX = points.reduce(0) { $0 + $1.x }
        let sumY = points.reduce(0) { $0 + $1.y }
        let avgX = sumX / CGFloat(points.count) * imageSize.width
        let avgY = (1 - sumY / CGFloat(points.count)) * imageSize.height
        
        return PointData(x: avgX, y: avgY)
    }
    
    private func convertPoint(_ point: CGPoint, imageSize: CGSize) -> PointData {
        return PointData(
            x: point.x * imageSize.width,
            y: (1 - point.y) * imageSize.height
        )
    }
}

// MARK: - Data Types

struct DetectedFace {
    var bounds: FaceBounds
    var landmarks: FaceLandmarksData?
    var contours: FaceContoursData?
    var smilingProbability: Double?
    var leftEyeOpenProbability: Double?
    var rightEyeOpenProbability: Double?
    var trackingId: Int?
    var headEulerAngleX: Double?
    var headEulerAngleY: Double?
    var headEulerAngleZ: Double?
    
    func toDictionary() -> [String: Any] {
        var dict: [String: Any] = [
            "bounds": [
                "x": bounds.x,
                "y": bounds.y,
                "width": bounds.width,
                "height": bounds.height
            ]
        ]
        
        if let landmarks = landmarks {
            dict["landmarks"] = landmarks.toDictionary()
        }
        
        if let contours = contours {
            dict["contours"] = contours.toDictionary()
        }
        
        if let smiling = smilingProbability {
            dict["smilingProbability"] = smiling
        }
        
        if let leftEye = leftEyeOpenProbability {
            dict["leftEyeOpenProbability"] = leftEye
        }
        
        if let rightEye = rightEyeOpenProbability {
            dict["rightEyeOpenProbability"] = rightEye
        }
        
        if let trackingId = trackingId {
            dict["trackingId"] = trackingId
        }
        
        if let angleY = headEulerAngleY {
            dict["headEulerAngleY"] = angleY
        }
        
        if let angleX = headEulerAngleX {
            dict["headEulerAngleX"] = angleX
        }
        
        if let angleZ = headEulerAngleZ {
            dict["headEulerAngleZ"] = angleZ
        }
        
        return dict
    }
}

struct FaceBounds {
    var x: CGFloat
    var y: CGFloat
    var width: CGFloat
    var height: CGFloat
}

struct PointData {
    var x: CGFloat
    var y: CGFloat
    
    func toDictionary() -> [String: Any] {
        return ["x": x, "y": y]
    }
}

struct FaceLandmarksData {
    var leftEye: PointData?
    var rightEye: PointData?
    var leftEar: PointData?
    var rightEar: PointData?
    var leftCheek: PointData?
    var rightCheek: PointData?
    var noseBase: PointData?
    var mouthLeft: PointData?
    var mouthRight: PointData?
    var mouthBottom: PointData?
    
    func toDictionary() -> [String: Any] {
        var dict: [String: Any] = [:]
        
        if let leftEye = leftEye { dict["leftEye"] = leftEye.toDictionary() }
        if let rightEye = rightEye { dict["rightEye"] = rightEye.toDictionary() }
        if let leftEar = leftEar { dict["leftEar"] = leftEar.toDictionary() }
        if let rightEar = rightEar { dict["rightEar"] = rightEar.toDictionary() }
        if let leftCheek = leftCheek { dict["leftCheek"] = leftCheek.toDictionary() }
        if let rightCheek = rightCheek { dict["rightCheek"] = rightCheek.toDictionary() }
        if let noseBase = noseBase { dict["noseBase"] = noseBase.toDictionary() }
        if let mouthLeft = mouthLeft { dict["mouthLeft"] = mouthLeft.toDictionary() }
        if let mouthRight = mouthRight { dict["mouthRight"] = mouthRight.toDictionary() }
        if let mouthBottom = mouthBottom { dict["mouthBottom"] = mouthBottom.toDictionary() }
        
        return dict
    }
}

struct FaceContoursData {
    var face: [[String: Any]]?
    var leftEyebrowTop: [[String: Any]]?
    var leftEyebrowBottom: [[String: Any]]?
    var rightEyebrowTop: [[String: Any]]?
    var rightEyebrowBottom: [[String: Any]]?
    var leftEye: [[String: Any]]?
    var rightEye: [[String: Any]]?
    
    func toDictionary() -> [String: Any] {
        var dict: [String: Any] = [:]
        
        if let face = face { dict["face"] = face }
        if let leftEyebrowTop = leftEyebrowTop { dict["leftEyebrowTop"] = leftEyebrowTop }
        if let leftEyebrowBottom = leftEyebrowBottom { dict["leftEyebrowBottom"] = leftEyebrowBottom }
        if let rightEyebrowTop = rightEyebrowTop { dict["rightEyebrowTop"] = rightEyebrowTop }
        if let rightEyebrowBottom = rightEyebrowBottom { dict["rightEyebrowBottom"] = rightEyebrowBottom }
        if let leftEye = leftEye { dict["leftEye"] = leftEye }
        if let rightEye = rightEye { dict["rightEye"] = rightEye }
        
        return dict
    }
}

